synphot/observation.py

```diff
@@ -694,3 +694,245 @@ def as_spectrum(self, binned=True, wavelengths=None):
+        of counts/second/pixel.
+        Default is 0 * (astropy.units.ct /
+        astropy.units.second / astropy.units.pixel)
+    readnoise : `~astropy.units.Quantity`, optional
+        Counts per pixel from the read noise with units of counts/pixel.
+        Default is 0 * astropy.units.ct / astropy.units.pixel.
+    gain : `~astropy.units.Quantity`, optional
+        Gain of the CCD with units of counts/ADU. Default is
+        1 * astropy.units.ct / astropy.units.adu such that the
+        contribution to the error due to the gain is assumed to be small.
+    ad_err : `~astropy.units.Quantity`, optional
+        An estimate of the 1 sigma error within the A/D converter with units of
+        adu/pixel. Default is set to
+        sqrt(0.289) * astropy.units.adu / astropy.units.pixel, where
+        sqrt(0.289) comes from the assumption that "for a charge level that is
+        half way between two output ADU steps, there is an equal chance that it
+        will be assigned to the lower or to the higher ADU value when converted
+        to a digital number" (text from footnote on page 56 of [1]_, see also
+        [2]_).
+
+    References
+    ----------
+    .. [1] Howell, S. B. 2000, *Handbook of CCD Astronomy* (Cambridge, UK:
+        Cambridge University Press)
+    .. [2] Merline, W. & Howell, S. B. *A Realistic Model for Point-sources
+        Imaged on Array Detectors: The Model and Initial Results*. ExA, 6:163
+        (1995)
+
+    Returns
+    -------
+    t : `~astropy.units.Quantity`
+        The exposure time needed (in seconds) to achieve the given signal
+        to noise ratio.
+    """
+    if countrate.unit in (u.electron / u.s, u.photon / u.s):
+        countrate = countrate.value * u.ct / u.s
+    elif countrate.unit != u.ct / u.s:
+        raise u.UnitsError('countrate must have units of (either '
+                           'astropy.units.ct, astropy.units.electron or '
+                           'astropy.units.photon) / astropy.units.s')
+
+    # necessary for units to work:
+    snr = snr * np.sqrt(1 * u.ct)
+    readnoise = _get_shotnoise(readnoise)
+    gain_err = _get_shotnoise(gain * ad_err)
+
+    # solve t with the quadratic equation (pg. 57 of Howell 2000)
+    A = countrate ** 2
+    B = (-1) * snr ** 2 * (countrate + npix * (background_rate +
+                                               darkcurrent_rate))
+    C = (-1) * snr ** 2 * npix * readnoise ** 2
+
+    t = (-B + np.sqrt(B ** 2 - 4 * A * C)) / (2 * A)
+
+    if gain_err.value > 1 or np.isfinite(n_background.value):
+        from scipy.optimize import fsolve
+        # solve t numerically
+        t = fsolve(_t_with_small_errs, t, args=(background_rate,
+                                                darkcurrent_rate,
+                                                gain_err, readnoise, countrate,
+                                                npix, n_background))
+        t = float(t) * u.s
```

Comment:
Ah, because `fsolve` returns an array. So to match the return type when `fsolve` isn't called, I convert it into a float. Though this does bring up a good point, that is, since `fsolve` returns the roots of the function (the function in this case being a quadratic), will `fsolve` ever return two solutions...? I'll have to think about this for a minute